#25 removed compatibility with older dovecot versions

I am running dovecot 2.3.7, as delivered with Ubuntu 20.04 LTS.
This would also hit Debian users on still-supported versions. Debian 10 provides dovecot 2.3.4, Debian 11 provides dovecot 2.3.13.

The fix for enabling compilation against newer dovecot versions broke backwards compatibility:

```
In file included from fts-elastic-plugin.c:11:
fts-elastic-plugin.h:42:52: error: macro "DOVECOT_PREREQ" passed 3 arguments, but takes just 2
   42 | #if defined(DOVECOT_PREREQ) && DOVECOT_PREREQ(2,3,0)
      |                                                    ^
In file included from /usr/include/dovecot/lib.h:27,
                 from fts-elastic-plugin.c:5:
/usr/include/dovecot/macros.h:226: note: macro "DOVECOT_PREREQ" defined here
  226 | #  define DOVECOT_PREREQ(maj, min) \
```

I don't know enough about C to know whether the macro signature could also be checked, to fix it in a way that works for both worlds.
It feels as if the Dovecot team shouldn't have made that change the way they did, at least not in a micro increment.

--- a/src/fts-elastic-plugin.c
+++ b/src/fts-elastic-plugin.c
@@ -75,6 +75,7 @@ fts_elastic_plugin_init_settings(struct mail_user *user,
     return 0;
 }
 
+#if defined(DOVECOT_PREREQ) && DOVECOT_PREREQ(2,3,17)
 static void fts_elastic_mail_user_deinit(struct mail_user *user)
 {
     struct fts_elastic_user *fuser = FTS_ELASTIC_USER_CONTEXT_REQUIRE(user);
@@ -82,13 +83,16 @@ static void fts_elastic_mail_user_deinit(struct mail_user *user)
     fts_mail_user_deinit(user);
     fuser->module_ctx.super.deinit(user);
 }
+#endif
 
 static void fts_elastic_mail_user_create(struct mail_user *user, const char *env)
 {
     FUNC_START();
-    struct mail_user_vfuncs *v = user->vlast;
     struct fts_elastic_user *fuser = NULL;
+#if defined(DOVECOT_PREREQ) && DOVECOT_PREREQ(2,3,17)
+    struct mail_user_vfuncs *v = user->vlast;
     const char *error;
+#endif
 
     /* validate our parameters */
     if (user == NULL || env == NULL) {
@@ -102,6 +106,7 @@ static void fts_elastic_mail_user_create(struct mail_user *user, const char *env
         return;
     }
 
+#if defined(DOVECOT_PREREQ) && DOVECOT_PREREQ(2,3,17)
     if (fts_mail_user_init(user, FALSE, &error) < 0) {
         i_error("fts_elastic: %s", error);
         return;
@@ -110,6 +115,7 @@ static void fts_elastic_mail_user_create(struct mail_user *user, const char *env
     fuser->module_ctx.super = *v;
     user->vlast = &fuser->module_ctx.super;
     v->deinit = fts_elastic_mail_user_deinit;
+#endif
 
     MODULE_CONTEXT_SET(user, fts_elastic_user_module, fuser);
     FUNC_END();

--- a/src/fts-elastic-plugin.h
+++ b/src/fts-elastic-plugin.h
@@ -39,6 +39,14 @@ void fts_elastic_plugin_deinit(void);
 
 #endif
 
+#if ((DOVECOT_VERSION_MAJOR << 24) + (DOVECOT_VERSION_MINOR << 16) + DOVECOT_VERSION_MICRO < ((2) << 24) + ((3) << 16) + (18))
+#undef DOVECOT_PREREQ
+#define DOVECOT_PREREQ(maj, min, micro) \
+       ((DOVECOT_VERSION_MAJOR << 24) + \
+        (DOVECOT_VERSION_MINOR << 16) + \
+        DOVECOT_VERSION_MICRO >= ((maj) << 24) + ((min) << 16) + (micro))
+#endif
+
 #if defined(DOVECOT_PREREQ) && DOVECOT_PREREQ(2,3,0)
 #else
 #   define str_append_max(str, data, size) str_append_n(str, data, size);

